crypto/modes/siv128.c

```diff
@@ -0,0 +1,365 @@
+#  if defined(__GNUC__) || defined(__clang__)
+__owur static ossl_inline uint64_t BSWAP8(uint64_t x)
+{
+    return __builtin_bswap64(x);
+}
+#  elif defined(_MSC_VER)
+__owur static ossl_inline uint64_t BSWAP8(uint64_t x)
+{
+    return _byteswap_uint64(x);
+}
+#  else
+__owur static ossl_inline uint32_t rotl(uint32_t x)
+{
+    return (x << 8) | (x >> 24);
+}
+__owur static ossl_inline uint32_t rotr(uint32_t x)
+{
+    return (x >> 8) | (x << 24);
+}
+__owur static ossl_inline uint64_t BSWAP8(uint64_t x)
+{
+    uint32_t high = (uint32_t)(x >> 32);
+    uint32_t low = (uint32_t)x;
+
+    high = (rotl(high) & 0x00ff00ff) | (rotr(high) & 0xff00ff00);
+    low = (rotl(low) & 0x00ff00ff) | (rotr(low) & 0xff00ff00);
+    return ((uint64_t)low) << 32 | (uint64_t)high;
+}
+#  endif
+# endif
+
+__owur static ossl_inline uint64_t siv128_getword(SIV_BLOCK const *b, size_t i)
+{
+    const union {
+        long one;
+        char little;
+    } is_endian = { 1 };
+
+    if (is_endian.little)
+        return BSWAP8(b->word[i]);
+    return b->word[i];
+}
+
+static ossl_inline void siv128_putword(SIV_BLOCK *b, size_t i, uint64_t x)
+{
+    const union {
+        long one;
+        char little;
+    } is_endian = { 1 };
+
+    if (is_endian.little)
+        b->word[i] = BSWAP8(x);
+    else
+        b->word[i] = x;
+}
+
+static ossl_inline void siv128_xorblock(SIV_BLOCK *x,
+                                     SIV_BLOCK const *y)
+{
+    x->word[0] ^= y->word[0];
+    x->word[1] ^= y->word[1];
+}
+
+/*
+ * Doubles |b|, which is 16 bytes representing an element
+ * of GF(2**128) modulo the irreducible polynomial
+ * x**128 + x**7 + x**2 + x + 1.
+ * Assumes two's-complement arithmetic
+ */
+static ossl_inline void siv128_dbl(SIV_BLOCK *b)
+{
+    uint64_t high = siv128_getword(b, 0);
+    uint64_t low = siv128_getword(b, 1);
+    uint64_t high_carry = high & (((uint64_t)1) << 63);
+    uint64_t low_carry = low & (((uint64_t)1) << 63);
+    int64_t low_mask = -((int64_t)(high_carry >> 63)) & 0x87;
+    uint64_t high_mask = low_carry >> 63;
+
+    high = (high << 1) | high_mask;
+    low = (low << 1) ^ (uint64_t)low_mask;
+    siv128_putword(b, 0, high);
+    siv128_putword(b, 1, low);
+}
+
+
+__owur static ossl_inline int siv128_do_s2v_p(SIV128_CONTEXT *ctx, SIV_BLOCK *out,
+                                              unsigned char const* in, size_t len)
+{
+    SIV_BLOCK t;
+    size_t out_len = sizeof(out->byte);
+
+    if (!CMAC_CTX_copy(ctx->cmac_ctx, ctx->cmac_ctx_init))
+        return 0;
+
+    if (len >= SIV_LEN) {
+        if (!CMAC_Update(ctx->cmac_ctx, in, len - SIV_LEN))
+            return 0;
+        memcpy(&t, in + (len-SIV_LEN), SIV_LEN);
+        siv128_xorblock(&t, &ctx->d);
+        if (!CMAC_Update(ctx->cmac_ctx, t.byte, SIV_LEN))
+            return 0;
+    } else {
+        size_t i;
+
+        memcpy(&t, in, len);
+        t.byte[len] = 0x80;
+        for (i = len + 1; i <SIV_LEN; i++)
+            t.byte[i] = 0;
```

Comment:
Wouldn't it be efficient to zero t prior memcpy? Contemporary compiler should be able to replace it with just pair of stores, but if you don't trust that it would you can `t.word[0] = t.word[1] = 0;`

Reply:
Just in case to clarify. Contemporary compiler should be able to replace `memset(&t,0,sizeof(t))` with just pair of stores, but if ...

Reply:
Assumes that SIV_LEN does not change... for the simple assignment. But a memcpy ought to be damned fast with something fixed and that size.

Reply:
please add a space after the "<": `i <SIV_LEN`